src/compiler/checker.ts

```diff
@@ -24699,6 +24700,20 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             }
         }
 
+        function inferFromMappedProperties(source: Type, target: MappedType) {
+            const properties = getPropertiesOfObjectType(source);

```

Comment:
I noticed this function is not inferring from index signatures, so the following example doesn't work:

```ts
type Selector<S, R> = (state: S) => R;

declare function createStructuredSelector<S, T>(selectors: {
  [K in keyof T]: Selector<S, T[K]>;
}): Selector<S, T>;

type State = { foo: number };

declare const otherSelector: { [x: string]: Selector<State, boolean> };
const otherResult = createStructuredSelector(otherSelector);
```
Ideally I think we want inference to work the same for the index signature case.

While looking around to see if we already have a function that you could use here, I found `forEachMappedTypePropertyKeyTypeAndIndexSignatureKeyType`, which seems to do what you'd want to do here of iterating through properties and index signatures and all that.

Reply:
good catch! although now I wonder if we shouldn't infer from The Holy Quadrinity.
https://github.dev/microsoft/TypeScript/blob/7e4592ea096b626274e7b20550f2f9cbad9a162f/src/compiler/checker.ts#L24848-L24851

Reply:
Indeed, you probably want mapped-type-aware forms of all of the structural breakdowns done in the following section. In fact, it might be easier to just remove the early `return` and make the tuple matching block, `inferFromProperties`, `inferFromSignatures`, and `inferFromIndexTypes` handle mapped types on the `target` side better. With the caveat that, for compatability's sake, you _may_ need a new priority for these inferences. But that's something you can only really decide after seeing how the change works on a bunch of real world code.